### PR TITLE
Improved CTD to galaxy conversion

### PR DESCRIPTION
**Pull request by bernt-matthias**
This is the result of my work for the auto generation of the Galaxy OpenMS wrappers. 

- adds testing
- hardcoded params now as json which allows to do a lot more now (e.g. blacklisting and overwriting of xml/ctd attributes)
- major overhaul of most parameters .. see also https://github.com/WorkflowConversion/CTDConverter/issues/48, https://github.com/WorkflowConversion/CTDConverter/issues/52, https://github.com/WorkflowConversion/CTDConverter/issues/53

superseeds https://github.com/WorkflowConversion/CTDConverter/pull/50 and https://github.com/WorkflowConversion/CTDConverter/pull/46

### CODE REVIEW

**Comment by bernt-matthias** [comment]
ping @Tomnl @bgruening

**Comment by bgruening** [comment]
@bernt-matthias can you run this and create a PR against galaxyp/openms ...?

Thanks a lot, this is awesome!

**Comment by bernt-matthias** [comment]
This is comming soon. Still fighting with a few failing tests (yeah we have tests :) ). 

this will also soon be ready https://github.com/galaxyproject/galaxy/pull/8109 .. maybe we should check if we can have this in 20.01

**Comment by bgruening** [comment]
Jupp, I think retargeting to 20.01 could work. That's a cool present! Thanks so much.

**Comment by Tomnl** [comment]
This is great! Thanks @bernt-matthias !

ping @RJMW

**Comment by RJMW** [comment]
Wow! - Thanks for the all the efforts @bernt-matthias

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+language: python
+python:
+  - "3.6"
+
+install:
+  - sudo apt-get update
+#   - if [[ "$TRAVIS_PYTHON_VERSION" == "2.7" ]]; then
```
Comment (jpfeuffer) [inline .travis.yml]: Still needed?

File: galaxy/macros.xml [hunk 2]
```diff
@@ -1,31 +1,122 @@
...
+  <xml name="stdio">
+    <stdio>
+      <regex match="Could not allocate metaspace" level="fatal_oom" description="Java memory Exception"/>
+      <regex match="Cannot create VM thread" level="fatal_oom" description="Java memory Exception"/>
+      <regex match="qUncompress: could not allocate enough memory to uncompress data" level="fatal_oom" description="Java memory Exception"/>
+    </stdio>
+  </xml>
+  <xml name="references">
+    <citations>
+      <citation type="doi">doi:10.1186/1471-2105-9-163</citation>
```
Comment (jpfeuffer) [inline galaxy/macros.xml]: Will this add OpenMS reference to everything that gets converted from CTD?
Maybe point out, that at least that part should be changed in the macros?

File: convert.py [hunk 3]
```diff
@@ -222,28 +222,28 @@ def main(argv=None):
...
-        utils.error("CTDConverter could not complete the requested operation.", 0)
-        utils.error("Reason: " + e.msg, 0)
+        utils.logger.error("CTDConverter could not complete the requested operation.", 0)
+        utils.logger.error("Reason: " + e.msg, 0)
         return 2
 
 
 def validate_and_prepare_common_arguments(args):
     # flatten lists of lists to a list containing elements
-    lists_to_flatten = ["input_files", "blacklisted_parameters"]
```
Comment (jpfeuffer) [inline convert.py]: Is this not needed or not supported anymore?
Or was it never needed here?

**Comment by mr-c** [comment]
Hello @bernt-matthias !

I'm working on the CWL export, and I see that you've made a lot of fixes that I have, or that I need. Do you anticipate being able to merge soon?

Or shall we peel off some of these fixes into other "bite-sized" PRs and merge them individually?

**Comment by bernt-matthias** [comment]
Hi @mr-c .. good point. I would be happy about comments and to get this merged. 

I would say that this is functional. We used this branch for the release of OpenMS Galaxy tools for version 2.5 and 2.6. 

But: 

- update of the docs would be nice. in particular parameter hardcoding etc.
- I guess also the failing tests should be considered

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+#   - conda install -c conda-forge lxml
+#   - conda install -c conda-forge ruamel.yaml 
+#   - conda install libxml2
+#   - cd $TRAVIS_BUILD_DIR
+#   - python setup.py install
+#   - conda install coverage green codecov
+
+#   - virtualenv .venv
+#   - . .venv/bin/activate
+  - pip install git+https://github.com/bernt-matthias/CTDopts@topic/no-1-2x
```
Comment (mr-c) [inline .travis.yml:42]: `WorkflowConversion:master is up to date with all commits from bernt-matthias:topic/no-1-2x.`

But the reverse is not true, and is the source of at least one of the failing tests

https://github.com/bernt-matthias/CTDopts/compare/topic/no-1-2x...WorkflowConversion:master#diff-a55c430e446e6696bb3f110ad5ce8607e83273a276ce6aaef8816a285b014503R50

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  - pip install lxml ruamel.yaml planemo 
+  - cd $TRAVIS_BUILD_DIR
+  - python setup.py install
+script:
+  - python setup.py test
+# TODO fail expected due to 
+# https://github.com/galaxyproject/galaxy/pull/9081 which will be in 20.01 and 
+# empty.xml (no inputs) could be removed and
```
Comment (mr-c) [inline .travis.yml:50]: Is this still the case? The mentioned PR was merged last year

**Comment by mr-c** [comment]
```
FAIL: test_galaxy_cli_bool_ctd (tests.test_galaxy_cli.GalaxyCliTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/WorkflowConversion/CTDConverter/tests/test_galaxy_cli.py", line 51, in test_galaxy_cli_bool_ctd
    self._compare_cli_output('bool')
  File "/home/travis/build/WorkflowConversion/CTDConverter/tests/test_galaxy_cli.py", line 48, in _compare_cli_output
    self.assertEqual(new_l[i].rstrip(), old_l[i].rstrip())
AssertionError: '#import os.path' != ''
- #import os.path
+ 
```

**Comment by mr-c** [comment]
PyTests pass!

Looks like some of the planemo-run tests fail. Shall we skip them and merge? 
```
ofile-mult (Test #1): failed
ofile-mult-typeparam (Test #1): failed
```

**Comment by bernt-matthias** [comment]
I will fix this this afternoon / tomorrow .. just need to remove a few test cases which are unreasonable (I think).

**Comment by mr-c** [comment]
so close, back to the error seen in https://github.com/WorkflowConversion/CTDConverter/pull/54#issuecomment-775098289

**Comment by bernt-matthias** [comment]
Just discussing with OpenMS developers that a change analogous to https://github.com/WorkflowConversion/CTDConverter/pull/54/commits/4912610f14998a1ad3a0dbc9cb150dc7ea51c13b might also be necessary for outputs :(

**Comment by bernt-matthias** [comment]
Wait, this is already happening :) .. But I was thinking about to change this slightly. 

Now for list input / outputs we create a folder `PARAMETER_NAME/GALAXY_UNIQUE_DATASET_ID/` and place the symlink with the name `label.ext` in there. 

I'm thinking about using an integers 0,1,2... instead of `GALAXY_UNIQUE_DATASET_ID` would be a bit nicer. 

And should be easy to implement.

**Comment by bernt-matthias** [comment]
Hah. Tests are passing. Give me a bit for the last step...

**Comment by bernt-matthias** [comment]
OK. If tests here and there https://github.com/galaxyproteomics/tools-galaxyp/pull/555 are passing its ready.

**Comment by bernt-matthias** [comment]
Looks good now. Last failing OpenMS tool is fixed. You may merge if you like.

For another automatic tool generation project (https://github.com/hexylena/argparse2tool/) I would like to reuse the ParameterHardcoder, i.e. 

https://github.com/WorkflowConversion/CTDConverter/blob/9247ce5d2397143a0683e179755be5e083b9d054/common/utils.py#L27

and

https://github.com/WorkflowConversion/CTDConverter/blob/9247ce5d2397143a0683e179755be5e083b9d054/common/utils.py#L215

Do you think its worth to factor this out into a separate project or should I just import CTDconverter?

Btw. `argparse2tool` can also generate CWL tools.

**Comment by mr-c** [comment]
> Looks good now. Last failing OpenMS tool is fixed. You may merge if you like.

Huzzah! I don't have write access to this repository yet, otherwise I would merge it :-)

> Do you think its worth to factor this out into a separate project or should I just import CTDconverter?

This isn't a javascript project, so I'd say that isn't enough for a separate package :-P

**Pull request merged**